### PR TITLE
capsules: ensure driver existence check

### LINKED ISSUE #3375: All drivers should return `Success` rather than `SuccessU32` for command 0
https://github.com/tock/tock/blob/master/doc/reference/trd104-syscalls.md#431-command-identifier-0 specifies that command 0 should always return `Success`. However, at least the following 4 drivers do not follow this direction:

- ADC https://github.com/tock/tock/blob/500caa5e882b7efe40158849fb489205f07c6a2e/capsules/src/adc.rs#L1149
- GPIO https://github.com/tock/tock/blob/500caa5e882b7efe40158849fb489205f07c6a2e/capsules/src/gpio.rs#L198
- LED https://github.com/tock/tock/blob/500caa5e882b7efe40158849fb489205f07c6a2e/capsules/src/led.rs#L95
- PWM https://github.com/tock/tock/pull/3367

I recall there being a desire for userspace to be able to build helper functions around command 0 with the assumption of a fixed Success variant for that command, so we should update these 4 drivers by adding a new command number to check the number of resources for that particular driver. This will require matching updates to `libtock-rs` and `libtock-c` for each

### PR DESCRIPTION
**Pull request by Samir-Rashid**
### Pull Request Overview

The driver liveliness check is defined in TRD104 4.3.1. I have updated all drivers in advance of Tock v3 (#3375). Make code comments and documentation consistent with regards to "driver existence check" terminology -- change as per #2908. "Driver check" falsely implies some sort of functionality checking.

### Testing Strategy

Unsure how to test. These are breaking changes.

### TODO or Help Wanted

There seems to be an established convention that command code 0 should return the number of things the driver is controlling. I moved those commands to the next available command code. Should these syscalls have some standardized code number?

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

**Comment by bradjc** [comment]
It's great that this is a comprehensive change because we do want cmd 0 to be a way to check if the driver exists. We don't want people starting from capsules that do not follow this convention so it's great to remove this broadly.

We can't change the cmd==0 for the drivers with a check in the "2.0" column here: https://github.com/tock/tock/blob/master/doc/syscalls/README.md As the comments indicate we can't change those until Tock 3.0.

For the others we can change, and I think we should.

> Should these syscalls have some standardized code number?

I think generally using "1" would make sense, but I don't think it's worth changing existing drivers to match. Just adding them as the next available number makes sense to me for existing drivers.

**Review by bradjc — CHANGES_REQUESTED** [review]
We can't break the stability guarantees.

**Comment by Samir-Rashid** [comment]
As I reverted stable drivers, I noticed there is no PWM driver [listed](https://github.com/tock/tock/blob/master/doc/syscalls/README.md#hardware-access). It should be there as it has an allocated driver number 10 (in hex?).

**Comment by hudson-ayers** [comment]
> As I reverted stable drivers, I noticed there is no PWM driver [listed](https://github.com/tock/tock/blob/master/doc/syscalls/README.md#hardware-access). It should be there as it has an allocated driver number 10 (in hex?).

Good catch, feel free to add that in this PR or another standalone one!

File: capsules/core/src/adc.rs [hunk 1]
```diff
@@ -1191,10 +1191,10 @@ impl<'a, A: hil::adc::Adc<'a> + hil::adc::AdcHighSpeed<'a>> SyscallDriver for Ad
             return CommandReturn::failure(ErrorCode::NOMEM);
         }
         match command_num {
-            // check if present
+            // Driver existence check
             // TODO(Tock 3.0): TRD104 specifies that Command 0 should return Success, not SuccessU32,
             // but this driver is unchanged since it has been stabilized. It will be brought into
-            // compliance as part of the next major release of Tock.
+            // compliance as part of the next major release of Tock. See #3613.
```
Comment (bradjc) [inline capsules/core/src/adc.rs:1197]: Can we point to #3375 instead? That is the tracking issue that will remain open for a possible 3.0 release.

File: capsules/core/src/gpio.rs [hunk 2]
```diff
@@ -198,8 +199,8 @@ impl<'a, IP: gpio::InterruptPin<'a>> SyscallDriver for GPIO<'a, IP> {
         let pins = self.pins.as_ref();
         let pin_index = data1;
         match command_num {
-            // number of pins
-            0 => CommandReturn::success_u32(pins.len() as u32),
+            // Check existence.
+            0 => CommandReturn::success(),
```
Comment (bradjc) [inline capsules/core/src/gpio.rs:203]: I think we need PRs open for libtock-c and libtock-rs to make this change.

https://github.com/tock/libtock-c/blob/master/libtock/gpio.c
https://github.com/tock/libtock-rs/blob/master/apis/gpio/src/lib.rs#L64

File: capsules/extra/src/analog_comparator.rs [hunk 3]
```diff
@@ -136,8 +137,8 @@ impl<'a, A: hil::analog_comparator::AnalogComparator<'a>> SyscallDriver
         processid: ProcessId,
     ) -> CommandReturn {
         if command_num == 0 {
-            // Handle this first as it should be returned unconditionally.
-            return CommandReturn::success_u32(self.channels.len() as u32);
+            // Handle unconditional driver existence check.
+            return CommandReturn::success();
```
Comment (bradjc) [inline capsules/extra/src/analog_comparator.rs:141]: Need pr: https://github.com/tock/libtock-c/blob/master/libtock/analog_comparator.c#L8

File: capsules/extra/src/gpio_async.rs [hunk 4]
```diff
@@ -169,9 +170,14 @@ impl<Port: hil::gpio_async::Port> SyscallDriver for GPIOAsync<'_, Port> {
         let other = (data >> 16) & 0xFFFF;
         let ports = self.ports.as_ref();
 
-        // Special case command 0; everything else results in a process-owned,
-        // split-phase call.
+        // Driver existence check.
         if command_number == 0 {
```
Comment (bradjc) [inline capsules/extra/src/gpio_async.rs:174]: We don't use this in libtock-c so we are good here.

File: capsules/extra/src/read_only_state.rs [hunk 5]
```diff
@@ -127,8 +128,11 @@ impl<'a, T: Time> SyscallDriver for ReadOnlyStateDriver<'a, T> {
         _processid: ProcessId,
     ) -> CommandReturn {
         match command_number {
-            // get version
-            0 => CommandReturn::success_u32(VERSION),
+            // Check existence
+            0 => CommandReturn::success(),
+
+            // Get version
+            1 => CommandReturn::success_u32(VERSION),
```
Comment (bradjc) [inline capsules/extra/src/read_only_state.rs:135]: https://github.com/tock/libtock-c/blob/master/libtock/read_only_state.c#L3

**Comment by Samir-Rashid** [comment]
Apology for delay. [Here](https://github.com/Samir-Rashid/tock/commit/ce1d50804fcf85a2779ec62503b50073838b24f8) is a diff of what I changed in case GitHub gets confused when I squash again. I made accompanying PRs https://github.com/tock/libtock-c/pull/349 and https://github.com/tock/libtock-rs/pull/517

File: doc/syscalls/00007_analog_comparator.md [hunk 6]
```diff
@@ -55,7 +55,7 @@ the index of the last AC module.
 
     **Returns**: `Ok(())` if starting interrupts was succesful.
 
-* ### Command number: `4`
+* ### Command number: `3`
```
Comment (ppannuto) [inline doc/syscalls/00007_analog_comparator.md:58]: Just noting for myself: this is not a change per se, but a fix of a documentation bug

**Review by ppannuto — APPROVED** [review]
I think this is looking pretty good. It's a lot of changes, but it's a nice consistency pass over a pretty old part of the kernel interface. It doesn't look like any stabilized interfaces are changed (though some of the documentation gets small updates in the name of consistency—which is good IMO).

**Review by bradjc — APPROVED** [review]

**Pull request merged**